packages/repository/src/repositories/kv.repository.bridge.ts

```diff
@@ -0,0 +1,69 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import * as legacy from 'loopback-datasource-juggler';
+
+import {Options, DataObject} from '../common-types';
+import {Model} from '../model';
+
+import {KVRepository, AsyncIterator, KVFilter} from './kv.repository';
+
+import {juggler, ensurePromise} from './legacy-juggler-bridge';
+
+/**
+ * An implementation of KVRepository based on legacy loopback-datasource-juggler
+ */
+export class DefaultKVRepository<T extends Model> implements KVRepository<T> {
+  kvModelClass: typeof juggler.KeyValueModel;
+
+  constructor(kvModelClass: typeof juggler.KeyValueModel) {
+    this.kvModelClass = kvModelClass;
+  }
+
+  delete(key: string, options?: Options): Promise<void> {
+    return ensurePromise(this.kvModelClass.delete(key, options));
+  }
+
+  deleteAll(options?: Options): Promise<void> {
+    return ensurePromise(this.kvModelClass.deleteAll(options));
+  }
+
+  get(key: string, options?: Options): Promise<T> {
+    const val = this.kvModelClass.get(key, options) as legacy.PromiseOrVoid<T>;
+    return ensurePromise<T>(val);
+  }
+
+  set(key: string, value: DataObject<T>, options?: Options): Promise<void> {
+    return ensurePromise<void>(this.kvModelClass.set(key, value, options));
+  }
+
+  expire(key: string, ttl: number, options?: Options): Promise<void> {
+    return ensurePromise<void>(this.kvModelClass.expire(key, ttl, options));
+  }
+
+  ttl(key: string, options?: Options): Promise<number> {
+    return ensurePromise<number>(this.kvModelClass.ttl(key, options));
+  }
+
+  keys(filter?: KVFilter, options?: Options): Promise<string[]> {
+    return ensurePromise<string[]>(this.kvModelClass.keys(filter, options));
+  }
+
+  iterateKeys(filter?: KVFilter, options?: Options): AsyncIterator<string> {
```

Comment:
In the previous pull request, I proposed to drop `keys(...): Promise<string[]>` and rename `iterateKeys` to `keys`. What are your arguments for keeping both flavors?

Cross-posting https://github.com/strongloop/loopback-next/pull/1527#discussion_r202748336:

Loading all keys into memory is a foot-gun that opens doors to a DoS attack where the datasource is queried with a pattern that matches too many keys.

The only reason why I added the method `keys(): string[]`, was to work around the limitations of LB 3.x - the way how it was directly exposing juggler methods via REST APIs and a lack of support for streaming responses.

We don't have that limitation in LB4 because REST API is exposed via custom Controller classes that can work around missing support for streaming responses in LB4.

I am proposing to remove `keys(): string[]` API from LB4 and provide only one method based on AsyncIterator. Ideally, this single method should be called `keys` instead of `iterateKeys` for simplicity.

Until our REST layer supports AsyncIterators, we can provide a helper function to convert AsyncIterator into a static array that people can use in their controllers.

Thoughts?

I am ok to leave this change out of this pull request if you prefer, as long as we make it happen before the first public release of KeyValue feature in LB4.

Reply:
I agree.